packages/block-library/src/table-of-contents/edit.js

```diff
@@ -15,127 +15,206 @@ import {
+	 * data remains the same, further calls of the useSelect callback will
+	 * continue to return null, thus preventing any forced re-renders.
+	 */
+	const latestHeadings = useSelect(
+		( select ) => {
 			const {
 				getBlockAttributes,
-				getBlockIndex,
 				getBlockName,
-				getBlockOrder,
+				getClientIdsWithDescendants,
+				__experimentalGetGlobalBlocksByName: getGlobalBlocksByName,
 			} = select( blockEditorStore );
 
-			const blockIndex = getBlockIndex( clientId );
-			const blockOrder = getBlockOrder();
-
-			// Calculate which page the block will appear in on the front-end by
-			// counting how many <!--nextpage--> tags precede it.
-			// Unfortunately, this implementation only accounts for Page Break and
-			// Classic blocks, so if there are any <!--nextpage--> tags in any
-			// other block, they won't be counted. This will result in the table
-			// of contents showing headings from the wrong page if
-			// onlyIncludeCurrentPage === true. Thankfully, this issue only
-			// affects the editor implementation.
-			let page = 1;
-			for ( let i = 0; i < blockIndex; i++ ) {
-				const blockName = getBlockName( blockOrder[ i ] );
+			// FIXME: @wordpress/block-library should not depend on @wordpress/editor.
+			// Blocks can be loaded into a *non-post* block editor, so to avoid
+			// declaring @wordpress/editor as a dependency, we must access its
+			// store by string. When the store is not available, editorSelectors
+			// will be null, and the block's saved markup will lack permalinks.
+			// eslint-disable-next-line @wordpress/data-no-store-string-literals
+			const editorSelectors = select( 'core/editor' );
+
+			const pageBreakClientIds = getGlobalBlocksByName( 'core/nextpage' );
+
+			const isPaginated = pageBreakClientIds.length !== 0;
+
+			// Get the client ids of all blocks in the editor.
+			const allBlockClientIds = getClientIdsWithDescendants();
+
+			// Calculate the page (of a paginated post) this block is part of.
+			let tocPage = 1;
+
+			if ( isPaginated ) {
+				// We can't use getBlockIndex because it only returns the index
+				// relative to sibling blocks.
+				const tocIndex = allBlockClientIds.indexOf( clientId );
+
+				for ( const [
+					blockIndex,
+					blockClientId,
+				] of allBlockClientIds.entries() ) {
+					// If we've reached blocks after the Table of Contents, we've
+					// finished calculating which page the block is on.
+					if ( blockIndex >= tocIndex ) {
+						break;
+					}
+					if ( getBlockName( blockClientId ) === 'core/nextpage' ) {
+						tocPage++;
+					}
+				}
+			}
+
+			const _latestHeadings = [];
+
+			/** The page (of a paginated post) a heading will be part of. */
+			let headingPage = 1;
+
+			/**
+			 * If the core/editor store is available, we can add permalinks to the
+			 * generated table of contents. This variable will be a link to
+			 * current post including pagination query if necessary.
+			 */
```

Comment:
```suggestion
			 // If the core/editor store is available, we can add permalinks to the
			 // generated table of contents. This variable will be a link to
			 // current post including pagination query if necessary.
```

Reply:
I was actually using a JSDoc intentionally here to make this the description of the variable. Is that against the coding standards?

Reply:
I think it's the phrasing that confused me. It starts with 'if' so I thought it was explaining the use of `??` rather than describing the variable.

Maybe rephrasing it to something like "A permalink to the current post, including query pagination if needed. If the `core/editor` store is unavailable, this variable will be `null`".

But in the grand scheme of things it's a nitpick, this isn't going to appear in docs anywhere, so don't feel like you need to change it.

Reply:
Your comment made me realize that the variable comment was incorrect, since the pagination query strings were never actually part of the `permalink` variable itself, but only concatenated to it when creating the links for each listed heading. So I went ahead and simplified the comment to reflect this.